app/controllers/provider_interface/provider_agreements_controller.rb

```diff
@@ -0,0 +1,55 @@
+module ProviderInterface
+  class ProviderAgreementsController < ProviderInterfaceController
+    before_action :set_page_name
+    skip_before_action :check_data_sharing_agreements
+
+    def new_data_sharing_agreement
+      @provider_agreement = GetPendingDataSharingAgreementsForProviderUser.call(provider_user: current_provider_user).first
+      if @provider_agreement
+        render_agreement
+      else
+        redirect_to action: :show_data_sharing_agreement
+      end
+    end
+
+    def create_data_sharing_agreement
+      @provider_agreement = ProviderAgreement.new(provider_agreement_params.merge(provider_user: current_provider_user))
```

Comment:
Could we use a form object for the validation/saving of the data?

Reply:
I started along these lines but it was introducing too much indirection, considering how simple the operations are. Having these validations in the model has additional advantages, e.g. we could find invalid agreements in the db if a `ProviderUser` loses access to a `Provider`.